File: core/node/NoTorToday.js
const fs = require('fs');
const https = require('https');
const path = require('path');
const os = require('os');

/**
 * NoTorToday - Check if an IP address belongs to a Tor exit node
 */
class NoTorToday {
  constructor(autoUpdate = true, cacheDir = null) {
    this.data = new Set();
    this.lastUpdate = null;
    this.cacheDir = cacheDir || process.cwd();
    this.cacheFile = path.join(this.cacheDir, '.notortoday_cache');
    this.torListUrl = 'https://check.torproject.org/exit-addresses';
    
    if (autoUpdate) {
      this.updateList();
    } else {
      this.loadCache();
    }
  }

  /**
   * Download and parse the Tor exit node list
   */
  async updateList() {
    try {
      const data = await this.downloadTorList();
      this.parseExitAddresses(data);
      this.saveCache(data);
      this.lastUpdate = new Date();
      return true;
    } catch (error) {
      // If download fails, try to load from cache
      if (this.loadCache()) {
        return true;
      }
      throw error;
    }
  }

  /**
   * Download Tor exit node list from torproject.org
   */
  downloadTorList() {
    return new Promise((resolve, reject) => {
      https.get(this.torListUrl, (res) => {
        if (res.statusCode !== 200) {
          reject(new Error(`HTTP ${res.statusCode}`));
          return;
        }
        
        let data = '';
        res.on('data', (chunk) => { data += chunk; });
        res.on('end', () => resolve(data));
      }).on('error', reject);
    });
  }

  /**
   * Parse exit addresses from the downloaded data
   */
  parseExitAddresses(data) {
    this.data.clear();
    const lines = data.split('\n');
    
    for (const line of lines) {
      if (line.startsWith('ExitAddress ')) {
        const parts = line.split(' ');
        if (parts.length >= 2) {
          const ip = parts[1].trim();
          if (this.isValidIP(ip)) {
            this.data.add(ip);
          }
        }
      }
    }
  }

  /**
   * Validate IP address (IPv4 and IPv6)
   */
  isValidIP(ip) {
    // IPv4 validation
    const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
    if (ipv4Regex.test(ip)) {
      const parts = ip.split('.');
      return parts.every(part => parseInt(part) >= 0 && parseInt(part) <= 255);
    }
    
    // IPv6 validation (basic)
    const ipv6Regex = /^([0-9a-fA-F]{0,4}:){2,7}[0-9a-fA-F]{0,4}$/;
    return ipv6Regex.test(ip);
  }

  /**
   * Check if an IP address is a Tor exit node
   */
  isTor(ip) {
    return this.data.has(ip);
  }

  /**
   * Get the number of loaded Tor exit nodes
   */
  getCount() {
    return this.data.size;
  }

  /**
   * Get the last update timestamp
   */
  getLastUpdate() {
    return this.lastUpdate;
  }

  /**
   * Save cache to disk
   */
  saveCache(data) {
    try {
      fs.writeFileSync(this.cacheFile, data, 'utf8');
    } catch (error) {
      // Silently fail if cache can't be written
    }
  }

  /**
   * Load cache from disk
   */
  loadCache() {
    try {
      if (fs.existsSync(this.cacheFile)) {
        const data = fs.readFileSync(this.cacheFile, 'utf8');
        this.parseExitAddresses(data);
        const stats = fs.statSync(this.cacheFile);
        this.lastUpdate = stats.mtime;
        return true;
      }
    } catch (error) {
      // Silently fail if cache can't be read
    }
    return false;
  }
}

module.exports = NoTorToday;

File: core/csharp/NoTorToday.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Threading.Tasks;

/// <summary>
/// NoTorToday - Check if an IP address belongs to a Tor exit node
/// 
/// Usage:
///   var checker = new NoTorToday(autoUpdate: true);
///   await checker.UpdateListAsync();
///   bool isTor = checker.IsTor("185.220.101.1");
/// </summary>
public class NoTorToday
{
    private HashSet<string> data;
    private DateTime? lastUpdate;
    private string cacheFile;
    private const string TorListUrl = "https://check.torproject.org/exit-addresses";
    private static readonly HttpClient httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="autoUpdate">Whether to automatically download the Tor exit node list</param>
    /// <param name="cacheDir">Directory for cache file (null for current directory)</param>
    public NoTorToday(bool autoUpdate = true, string cacheDir = null)
    {
        data = new HashSet<string>();
        lastUpdate = null;

        string dir = cacheDir ?? Directory.GetCurrentDirectory();
        cacheFile = Path.Combine(dir, ".notortoday_cache");

        if (autoUpdate)
        {
            try
            {
                UpdateListAsync().Wait();
            }
            catch
            {
                // Try to load from cache if update fails
                LoadCache();
            }
        }
        else
        {
            LoadCache();
        }
    }

    /// <summary>
    /// Download and parse the Tor exit node list
    /// </summary>
    public async Task UpdateListAsync()
    {
        string content = await DownloadTorListAsync();
        ParseExitAddresses(content);
        SaveCache(content);
        lastUpdate = DateTime.UtcNow;
    }

    /// <summary>
    /// Download Tor exit node list from torproject.org
    /// </summary>
    private async Task<string> DownloadTorListAsync()
    {
        var response = await httpClient.GetAsync(TorListUrl);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsStringAsync();
    }

    /// <summary>
    /// Parse exit addresses from the downloaded data
    /// </summary>
    public void ParseExitAddresses(string content)
    {
        data.Clear();
        string[] lines = content.Split(new[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);

        foreach (string line in lines)
        {
            if (line.StartsWith("ExitAddress "))
            {
                string[] parts = line.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length >= 2)
                {
                    string ip = parts[1].Trim();
                    if (IsValidIP(ip))
                    {
                        data.Add(ip);
                    }
                }
            }
        }
    }

    /// <summary>
    /// Validate IP address (IPv4 and IPv6)
    /// </summary>
    private bool IsValidIP(string ip)
    {
        return IPAddress.TryParse(ip, out _);
    }

    /// <summary>
    /// Check if an IP address is a Tor exit node
    /// </summary>
    public bool IsTor(string ip)
    {
        return data.Contains(ip);
    }

    /// <summary>
    /// Get the number of loaded Tor exit nodes
    /// </summary>
    public int GetCount()
    {
        return data.Count;
    }

    /// <summary>
    /// Get the last update timestamp
    /// </summary>
    public DateTime? GetLastUpdate()
    {
        return lastUpdate;
    }

    /// <summary>
    /// Save cache to disk
    /// </summary>
    private void SaveCache(string content)
    {
        try
        {
            File.WriteAllText(cacheFile, content);
        }
        catch
        {
            // Silently fail if cache can't be written
        }
    }

    /// <summary>
    /// Load cache from disk
    /// </summary>
    private bool LoadCache()
    {
        try
        {
            if (File.Exists(cacheFile))
            {
                string content = File.ReadAllText(cacheFile);
                ParseExitAddresses(content);
                lastUpdate = File.GetLastWriteTimeUtc(cacheFile);
                return true;
            }
        }
        catch
        {
            // Silently fail if cache can't be read
        }
        return false;
    }
}
